code/audits/audit-service.md

```diff
@@ -0,0 +1,79 @@
+
+>An assumption in our storage system is that most storage nodes are reasonably well-behaved, and most data is stored faithfully. As long as that assumption holds, Reed-Solomon is able to detect errors and even correct them, via mechanisms such as the Berlekamp-Welch error correction algorithm [37,65].
+
+Excerpt from [Storj Whitepaper](https://storj.io/storjv3.pdf) Section 4.13, where more information and a higher-level description of the audit service can be found.
+
+This document describes the following:
+- Audit Service Components
+- Overview of an Audit
+- Stripe Selection
+- Verification
+- Containment Mode
+- Reverification
+
+# Components
+
+- Cursor - keeps track of auditing location as while iterating through the metainfoDB (or pointerDB)
+- Verifier - verifies the correctness of a given stripe (the bulk of "auditing" occurs here)
+- Service - coordinates the Cursor and Verifier to run the audit process continuously
+- Reporter - collates audit results and records the stats in the overlay cache
+
+# Overview of an Audit
+
+The audit cursor uses the metainfoDB to select a random stripe from a random remote segment. The audit verifier downloads all erasure shares generated for that stripe. The verifier needs at least the Reed-Solomon minimum required number of pieces (K) in order to verify the stripe's correctness. The verifier uses the Berlekamp-Welch algorithm to determine if any shares have been damaged or altered. If so, the verifier marks the culpable nodes as having failed the audit. If the share has not been altered, then the verifier will record the responsible nodes' audit successes. If the node appears to be offline at the time of downloading, the verifier will record the node as offline.
+
+In a special case, if a node appears to be online but does not send the requested share data to the satellite, then the verifier will mark the node as `contained`, and a pending audit will be entered in the ContainmentDB. The contained node will have three more opportunities (the configurable `MaxReverifyCount`) to produce the correct data, before it's automatically marked as an audit failure.
+
+Steps within Service:
+1. The audit service uses the cursor to select a random stripe.
+2. The verifier calls `Reverify` on the stripe in order to verify the correctness of data held by contained nodes.
+3. `Reverify` returns a report detailing if any contained nodes were found to pass the audit, fail the audit, or remain contained. The reporter records these results to the overlay cache.
+4. The service then calls `Verify` on the stripe, which will skip any nodes that were `reverified` already.
+5. `Verify` also returns a report of the audit results, which are then recorded to the overlay cache.
+
+# Stripe Selection
+
+1. The cursor gets a list of pointers from the metainfoDB, keeping track of the last path listed.
+2. The cursor finds a valid pointer, by filtering out pointers that are expired, inline, or have a segment size of 0.
```

Comment:
Out of curiosity, when is the size of segment 0?

Reply:
that's a great question and I don't remember the answer! I just know we had issues when we didn't check for that. I think @mobyvb might know?

Reply:
I think this might be the case for bucket pointers.